File: perf/benchmarks/apps/levenshtein/levenshtein.py
# levenshtein — edit-distance DP application benchmark.
#
# Build two deterministic strings A and B (length L each, characters drawn
# from 'a'..'h' via independent LCG streams) and compute their Levenshtein
# edit distance using a 1-D rolling DP. Argument: L (default 1500).

import sys
import time


def make_str(seed: int, n: int) -> str:
    out = []
    for _ in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        out.append(chr(ord('a') + ((seed >> 16) & 7)))
    return "".join(out)


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 1500
    a = make_str(1, n)
    b = make_str(7, n)

    t0 = time.perf_counter_ns()
    prev = list(range(n + 1))
    cur = [0] * (n + 1)
    for i in range(1, n + 1):
        cur[0] = i
        ai = a[i - 1]
        for j in range(1, n + 1):
            cost = 0 if ai == b[j - 1] else 1
            v = prev[j - 1] + cost
            d = cur[j - 1] + 1
            if d < v:
                v = d
            d = prev[j] + 1
            if d < v:
                v = d
            cur[j] = v
        prev, cur = cur, prev
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{prev[n]}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/apps/levenshtein/levenshtein.cpp
// levenshtein — edit-distance DP application benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

static std::string make_str(int64_t seed, int n) {
    std::string s;
    s.reserve(n);
    for (int i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        s.push_back(static_cast<char>('a' + ((seed >> 16) & 7)));
    }
    return s;
}

int main(int argc, char** argv) {
    int n = (argc > 1) ? std::atoi(argv[1]) : 1500;
    std::string a = make_str(1, n);
    std::string b = make_str(7, n);

    auto t0 = std::chrono::steady_clock::now();
    std::vector<int> prev(n + 1), cur(n + 1);
    for (int j = 0; j <= n; ++j) prev[j] = j;
    for (int i = 1; i <= n; ++i) {
        cur[0] = i;
        char ai = a[i - 1];
        for (int j = 1; j <= n; ++j) {
            int cost = (ai == b[j - 1]) ? 0 : 1;
            int v = prev[j - 1] + cost;
            int d = cur[j - 1] + 1;
            if (d < v) v = d;
            d = prev[j] + 1;
            if (d < v) v = d;
            cur[j] = v;
        }
        std::swap(prev, cur);
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%d\n", prev[n]);
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
